## IT2313 - Programming for Data Science
## Practical 5 - Question 1 to 11
## Data Manipulation with Pandas (Part 1)

### In this lesson, we will continue to practise our Pandas knowledge. 
- Importing Data into Pandas
- Changing the Values in DataFrame
- DataFrameData Exploration, Sorting and Filtering
- Slicing and Subsetting with .loc and .iloc
- Aggregating Methods

#### We need to import the neccessary libraries

In [ ]:
import numpy as np
import pandas as pd

#### Q1 - Let use the following Dictionary known as "dict" and load it into the Pandas DataFrame known as brics

In [2]:
dict = { "country":["Brazil", "Russia", "India", "China", "South Africa"],
         "capital":["Brasilia", "Moscow", "New Delhi", "Beijing", "Pretoria"],
         "area":[8.516, 17.10, 3.286, 9.597, 1.221],
         "population":[200.4, 143.5, 1252, 1357, 52.98] }
dict

{'country': ['Brazil', 'Russia', 'India', 'China', 'South Africa'],
 'capital': ['Brasilia', 'Moscow', 'New Delhi', 'Beijing', 'Pretoria'],
 'area': [8.516, 17.1, 3.286, 9.597, 1.221],
 'population': [200.4, 143.5, 1252, 1357, 52.98]}

In [4]:
# Type your Q1 answer here
brics=pd.DataFrame(dict)
brics


country    capital    area  population
0        Brazil   Brasilia   8.516      200.40
1        Russia     Moscow  17.100      143.50
2         India  New Delhi   3.286     1252.00
3         China    Beijing   9.597     1357.00
4  South Africa   Pretoria   1.221       52.98

#### Q2 - Change the Dataframe Index to : ["BR", "RU", "IN", "CH", "SA"]

In [7]:
# Type your Q2 answer here

index=["BR", "RU", "IN", "CH", "SA"]
brics=pd.DataFrame(dict,index=index)
brics

country    capital    area  population
BR        Brazil   Brasilia   8.516      200.40
RU        Russia     Moscow  17.100      143.50
IN         India  New Delhi   3.286     1252.00
CH         China    Beijing   9.597     1357.00
SA  South Africa   Pretoria   1.221       52.98

#### Q3 - Let export your DataFrame into a csv file know as brics.csv. Use Excel to open the csv file and compare with the DataFrame

In [8]:
# Type your Q3 answer here
# Export DataFrame to CSV
brics.to_csv('brics.csv')


#### Q4 - Let select the information for China using "loc" and display in Pandas Series and Pandas DataFrame

In [10]:
# Type your Q4 answer here for Pandas Series using "loc"
brics.loc["China":"China"]
 

KeyError: 'China'

In [7]:
# Type your Q4 answer here for Pandas DataFrame using "loc"



country  capital   area  population
CH   China  Beijing  9.597      1357.0

#### Q5 - Do select and display Row 2 to 4 of the DataFrame as shown

|    | country | capital   | area   | population |
|----|---------|-----------|--------|------------|
| RU | Russia  | Moscow    | 17.100 | 143.5      |
| IN | India   | New Delhi | 3.286  | 1252.0     |
| CH | China   | Beijing   | 9.597  | 1357.0     |

In [13]:
# Type your Q5 answer here, there are many ways to achieve the selection.
brics.iloc[1:4]


country    capital    area  population
RU  Russia     Moscow  17.100       143.5
IN   India  New Delhi   3.286      1252.0
CH   China    Beijing   9.597      1357.0

#### Q6 - Let use "loc" to select the and display the following:
|    | country | capital   | area   | population |
|----|---------|-----------|--------|------------|
| RU | Russia  | Moscow    | 17.100 | 143.5      |
| IN | India   | New Delhi | 3.286  | 1252.0     |

In [14]:
# Type your Q6 answer here
brics.loc["RU":"IN"]


country    capital    area  population
RU  Russia     Moscow  17.100       143.5
IN   India  New Delhi   3.286      1252.0

#### Q7 - Let use "loc" to select the and display the following:
|    | country | capital   |
|----|---------|-----------|
| RU | Russia  | Moscow    |
| IN | India   | New Delhi |

In [10]:
# Type your Q7 answer here



country    capital
RU  Russia     Moscow
IN   India  New Delhi

#### Q8 - Let use "loc" to select the and display the following:
|    | country       | capital           |
|----|---------------|-------------------|
| BR | Brazil        | Brasilia          |
| RU | Russia        | Moscow            |
| IN | India         | New Delhi         |
| CH | China         | Beijing           |
| SA | South Africia | Pretoria          |

In [11]:
# Type your Q8 answer here



country    capital
BR        Brazil   Brasilia
RU        Russia     Moscow
IN         India  New Delhi
CH         China    Beijing
SA  South Africa   Pretoria

#### Q9 - Let use "iloc" to select the and display the following:

|    | country | capital   | area   | population |
|----|---------|-----------|--------|------------|
| RU | Russia  | Moscow    | 17.100 | 143.5      |
| IN | India   | New Delhi | 3.286  | 1252.0     |
| CH | China   | Beijing   | 9.597  | 1357.0     |

In [12]:
# Type your Q9 answer here



country    capital    area  population
RU  Russia     Moscow  17.100       143.5
IN   India  New Delhi   3.286      1252.0
CH   China    Beijing   9.597      1357.0

#### Q10 - Let use "iloc" to select the and display the following:

|    | country | capital   |
|----|---------|-----------|
| RU | Russia  | Moscow    |
| IN | India   | New Delhi |
| CH | China   | Beijing   |

In [13]:
# Type your Q10 answer here



country    capital
RU  Russia     Moscow
IN   India  New Delhi
CH   China    Beijing

##### Q11 - Let import dataset from "temperatures.csv" into DataFrame known as temperatures

- a. Display the top 5 and bottom 5 rows of the DataFrame
- b. Display the number of rows and columns in this DataFrame
- c. Drop the column "s/n" from the DataFrame
- d. Generate descriptive statistics of the DataFrame's numerical columns
- e. Setting the index of the DataFrame to be a multi-level index using the columns "country" and "city" and display the Top 5 rows
- f. Sort the DataFrame based on Indexes in _ascending_ order and display the Top 5 rows and display the top 5 rows.
- g. Select rows starting from the combination ("India", "Hyderabad") up to and including the combination ("Iraq", "Baghdad").
- h. Select the rows which filter the rows in the DataFrame based on the specified date range. From 1 Jan 2010 to 31 Jan 2010

In [14]:
# Import your Dataset from "temperatures.csv" into DataFrame known as temperatures
temperatures=pd.read_csv('temperatures.csv')


In [15]:
# Type your Q11 (a) answer here



s/n      date     city        country  avg_temp_c
0    0  1/1/2000  Abidjan  Côte D'Ivoire      27.293
1    1  1/2/2000  Abidjan  Côte D'Ivoire      27.685
2    2  1/3/2000  Abidjan  Côte D'Ivoire      29.061
3    3  1/4/2000  Abidjan  Côte D'Ivoire      28.162
4    4  1/5/2000  Abidjan  Côte D'Ivoire      27.547

In [16]:
# Type your Q11 (a) answer here



s/n      date  city country  avg_temp_c
16495  16495  1/5/2013  Xian   China      18.979
16496  16496  1/6/2013  Xian   China      23.522
16497  16497  1/7/2013  Xian   China      25.251
16498  16498  1/8/2013  Xian   China      24.528
16499  16499  1/9/2013  Xian   China         NaN

In [17]:
# Type your Q11 (b) answer here



(16500, 5)

In [18]:
# Type your Q11(c) answer here



In [19]:
# Type your Q11(d) answer here



count       mean       std     min    25%     50%      75%  \
avg_temp_c  16407.0  19.732877  9.579635 -23.495  14.87  22.119  27.0015   

               max  
avg_temp_c  38.283

In [20]:
# Type your Q11(e) answer here



date  avg_temp_c
country       city                         
Côte D'Ivoire Abidjan  1/1/2000      27.293
              Abidjan  1/2/2000      27.685
              Abidjan  1/3/2000      29.061
              Abidjan  1/4/2000      28.162
              Abidjan  1/5/2000      27.547

In [21]:
# Type your Q11(f) answer here



date  avg_temp_c
country     city                        
Afghanistan Kabul   1/1/2000       3.326
            Kabul   1/2/2000       3.454
            Kabul   1/3/2000       9.612
            Kabul   1/4/2000      17.925
            Kabul   1/5/2000      24.658
...                      ...         ...
Zimbabwe    Harare  1/5/2013      18.298
            Harare  1/6/2013      17.020
            Harare  1/7/2013      16.299
            Harare  1/8/2013      19.232
            Harare  1/9/2013         NaN

[16500 rows x 2 columns]

In [22]:
# Type your Q11(g) answer here



date  avg_temp_c
country city                           
India   Hyderabad  1/1/2000      23.779
        Hyderabad  1/2/2000      25.826
        Hyderabad  1/3/2000      28.821
        Hyderabad  1/4/2000      32.698
        Hyderabad  1/5/2000      32.438
...                     ...         ...
Iraq    Baghdad    1/5/2013      28.673
        Baghdad    1/6/2013      33.803
        Baghdad    1/7/2013      36.392
        Baghdad    1/8/2013      35.463
        Baghdad    1/9/2013         NaN

[2145 rows x 2 columns]

In [23]:
# Type your Q11(h) answer here



date     city        country  avg_temp_c
120   2010-01-01  Abidjan  Côte D'Ivoire      28.270
121   2010-01-02  Abidjan  Côte D'Ivoire      29.262
122   2010-01-03  Abidjan  Côte D'Ivoire      29.596
123   2010-01-04  Abidjan  Côte D'Ivoire      29.068
124   2010-01-05  Abidjan  Côte D'Ivoire      28.258
...          ...      ...            ...         ...
16462 2010-01-08     Xian          China      23.434
16463 2010-01-09     Xian          China      19.630
16464 2010-01-10     Xian          China      12.292
16465 2010-01-11     Xian          China       6.742
16466 2010-01-12     Xian          China       0.845

[1200 rows x 4 columns]